### ISSUE TITLE
API: Streaming and downloading files is very slow

### ISSUE THREAD

**Issue by swingstate** [photoprism/photoprism#763]
I have now approx 40K photo and 4k videos in photoprism. All videos are .mp4 with max 4K resolution.

When I stream or download a file through the web interface of photoprism the transmission is very slow. 

Downloads: 1,6 - 1,9 Megabyte/sec
Stream: 15 - 20 Mbit/sec

I run photoprism on Docker. When I browse the same media folder, which photoprism uses too, with another container app (Cloud commander) and download a large file I get between 30 - 40 Megabyte / sec. Here my Wifi is only limiting.

in consequence of this low speed videos are not being started, they buffer for long time or they stop every few seconds. The 4K footage I use has a bitrate of 60 - 80 Mbit/sec

Any advise why even downloads of files (not streaming) are so slow and how this can be addressed?

The network is a Gigabit network with 700 -1200 Mbit on wifi. The disk subsystem is  Raid 10 PCIe 4 NVME based with Read and write speeds of 3 - 4GB/sec.

**Comment by lastzero** [comment]
Does this happen while indexing or otherwise high system load? I can only guess, don't have any such issues. Mainly depends on the underlying operating system and storage. Docker on macOS is known to have bad I/O for example.

**Comment by lastzero** [comment]
Note that you may use WebDAV for syncing and bulk downloads. Is this slow as well?

**Comment by swingstate** [comment]
The system is a 12 core / 64 GB DDR4, idling. No indexing running. OS underlying is a linux. Basically this container makes 2MB/sec, another up to 20x more.

**Comment by swingstate** [comment]
will destroy the container and rebuild it.

**Comment by lastzero** [comment]
Don't have 4k videos for testing. In theory, our specific streaming implementation may cause cause performance issues with very large videos. Not sure why this would happen in practice. Give us some time to look into this if you're sure this is a general issue.

**Comment by swingstate** [comment]
rebuilding the container with same db & media folder didn't improve.
I am creating a new container on same host with empty db and media folder and will test and post here.

**Comment by lastzero** [comment]
What specific API endpoint and client do you use for testing? Cloud commander probably doesn't do transcoding and streaming? Same issues with WebDAV?

**Comment by swingstate** [comment]
my client is a MacBook with Safari. Will test webdav too.

**Comment by lastzero** [comment]
This is where we return the file contents in our API - we use the standard HTTP server function for this, which should also support streaming: https://github.com/photoprism/photoprism/blob/develop/internal/api/video.go#L95

[Documentation](https://gowalker.org/github.com/gin-gonic/gin#Context_File) says:
> File writes the specified file into the body stream in a efficient way 

Note that we have to perform auth and format checks before downloading can start. That's different from a pure download server that doesn't perform any checks or transcoding.

**Comment by swingstate** [comment]
webdav upload seems to be much faster, using finder on MacOS. I am testing with a brand new container atm. Web performance, especially downloads are very slow with <1 MB/sec.
My container is running on UnRaid, where @ich777 is maintaining a repo. Eventually he can help to pin this down?

**Comment by swingstate** [comment]
sometimes the webdav upload doesn't even start.

**Comment by lastzero** [comment]
Also never had issues with uploading. You use macOS as client, start an upload and nothing happens? No error or success message whatsoever?

**Comment by swingstate** [comment]
See edited msg above.

**Comment by lastzero** [comment]
Http streaming should explicitly prevent the need for a fast upfront download, so I'm not 100% what "slow download" means in this context as that is the goal of streaming. Of course not when the video is not playable at the end! It's already late, can take a look later this week.

**Comment by swingstate** [comment]
download means when I click the edit button, where can update the meta data for files. 

<img width="1114" alt="Bildschirmfoto 2020-12-28 um 23 42 31" src="https://user-images.githubusercontent.com/73672554/103247225-739b9500-4966-11eb-9499-3f298e24678f.png">

**Comment by lastzero** [comment]
Note that our solution is not designed or optimized for very large video files. It's a photo app after all. If there is an easy fix, we'll do it.

**Comment by swingstate** [comment]
Looks like this could be the reason. Tested import and download with a 54MB photo and speeds where around 6MB/sec.

**Comment by lastzero** [comment]
Just a goal not a reason. Huge files are not part of our standard tests and optimizations. When the goal is to download a file very fast, you simply read and instantly return every byte, but then you can't stream because the server can't just jump to specific positions. It's a different goal and implementation, but I still see no reason why it should be super slow as you described. Web upload might be "limited" by the way we handle it with JavaScript, that is also optimized for many smaller files.

**Comment by lastzero** [comment]
Use an app like PhotoSync for fast syncing via WebDAV. Using the Web upload should be an exception when there is no other way, like on our demo.

**Comment by swingstate** [comment]
I use PhotoSync to Upload to Nextcloud first, review and then import to Photoprism - since there is no delete function in the UI yet I want to be sure want I add to the system.

**Comment by swingstate** [comment]
If you want to position photoprism for pictures and not / less for high res videos, I have no issue with that. I'll move my videos to another solution then. 
Even though I think the support for high res videos (every camera can do 4k today) would benefit photoprism as a "digital media management platform".

**Comment by lastzero** [comment]
I think it's important not to be all over the place, the focus of our PHOTO app should be clear :) 

Nextcloud should be much slower than PhotoPrism for syncing with PhotoSync. Deleting will be added soon, but you can also review files via WebDAV or any other app directly in the originals folder before starting to index.

**Comment by lastzero** [comment]
Eventually we might implement a complete video and digital media management solution that may also completely replace Nextcloud. It's just out of scope for a bug ticket like this while we're focusing on a stable release.

**Comment by swingstate** [comment]
Nextcloud isn't anything like a media management solution and I wouldn't use it if there where better solution for file mgmt. So far I only use it as entry point for files, but will reconsider your tip to webdav into PP. 
Indexing made be some trouble see #768.

Bottomline, I'll move my videos to another solution then. Focusing on "just" photo management is already complex :)

**Comment by swingstate** [comment]
I might start from scratch with a fresh db and container. Can I re-import my current originals folder to a new container and will the exif data be shown like in the current container? especially dates and gps data?

I want to split videos and photos and not have a photoprism container with any kind of video related lingering objects in the file system or db.

**Comment by lastzero** [comment]
You have to reindex when starting with a fresh database. Backups are in the Yaml sidecar files that you may choose to keep. Same for albums. Can be found in the storage folder.

**Comment by swingstate** [comment]
ok, but the files in the originals folder still contain all exif data or is there anything being changed during import (because I already imported the files once)?

**Comment by lastzero** [comment]
Exifdata is part of the image files, we don't modify them - just getting renamed during import. Make sure to use MariaDB for your new instance and not Sqlite. Major performance difference.

**Comment by lastzero** [comment]
Tested downloading videos from our public demo, performance only limited by the Internet connection:

```
wget https://demo.photoprism.org/api/v1/videos/81aefaee21a2882102a5cac0532f265965c7c679/public/avc
Resolving demo.photoprism.org (demo.photoprism.org)... 45.132.246.223
Connecting to demo.photoprism.org (demo.photoprism.org)|45.132.246.223|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 22251038 (21M) [video/mp4]
Saving to: ‘avc’

avc                                   100%[=========>]  21,22M  9,31MB/s    in 2,3s  

```

**Comment by lastzero** [comment]
**Wow!**

Local test is even faster, only limited by the disk and ethernet connection as it seems (no SSD):

```
wget http://myserver/api/v1/videos/1d61163e07140d72e1a68ab7804d6653f6d89c94/public/avc
Resolving neon (neon)... 192.168.178.2
Connecting to neon (neon)|192.168.178.2|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 711610493 (679M) [video/mp4]

678,64M  94,7MB/s    in 7,0s    
```

Streaming parts of a video file in a browser seems to work perfectly too:

**Request**

```
Accept: */*
Accept-Encoding: identity;q=1, *;q=0
Accept-Language: en-US,en;q=0.9,de-DE;q=0.8,de;q=0.7
Connection: keep-alive
DNT: 1
If-Range: Tue, 29 Apr 2014 09:26:12 GMT
Range: bytes=30572544-33392127
```

**Response**

```
Accept-Ranges: bytes
Content-Length: 681037949
Content-Range: bytes 30572544-711610492/711610493
Content-Type: video/mp4; codecs="avc1"
Date: Wed, 30 Dec 2020 18:37:01 GMT
Last-Modified: Tue, 29 Apr 2014 09:26:12 GMT
```

If you want me to perform one last test with a super huge super high res video, please send a download link to hello@photoprism.org 👍

**Comment by lastzero** [comment]
@swingstate Do you run PhotoPrism behind a reverse proxy like nginx?

**Comment by swingstate** [comment]
nope. at the moment it is only available in the local lan. With the new container and a fresh index of all 40K photos I get 4-6MB/sec vs. at the same time 30 - 35MB/sec through cloud commander (Here my Wifi is limiting, I get approx. 500Mbit Brutto bandwidth, tested with OpenSpeedtest in a 3rd container) . Everything is same when I test. The only difference is the container I use to access the raw data on the disk.

**Comment by swingstate** [comment]
I don't use Wget to test, just downloading through the Safari browser.

**Comment by lastzero** [comment]
Safari was downloading so fast, I couldn't even take a screenshot (no, Apple is somehow hiding this info from me - it's almost midnight I don't want to try more settings before I can see the actual download happening in front of my eyes).

However - good news - I managed to record a video of Chrome downloading a .mov file at excellent speed:

https://user-images.githubusercontent.com/301686/103384835-2cdda480-4af8-11eb-842a-6d3697840ea1.mov

**How to proceed...**

Let's close this issue and you send us a video file which is very slow for you plus a video showing how slow your browser is downloading it. Our email for this is hello@photoprism.org. In addition, we need your server specs and a list of all software you use (like proxy servers, Docker version, RAID,... whatever seems relevant). Is that a deal?

**Comment by swingstate** [comment]
deal
